File: Project Nothing/Framework/src/framework/network/TcpConnection.cpp
#include "stdafx.h"
#include "framework/network/NetBridge.h"
#include "framework/network/NetProtocol.h"
#include "framework/network/TcpSession.h"
#include "framework/manager/ProtocolManager.h"
#include "framework/network/TcpConnection.h"

CTcpConnection::CTcpConnection (std::shared_ptr<CTcpSession> _pkTcpSession)
	: m_pkTcpSession (_pkTcpSession)
	, m_kRemoteEndPoint (_pkTcpSession->GetRemoteEndpoint ())
{
}

CTcpConnection::~CTcpConnection ()
{
}

void CTcpConnection::Init ()
{
	m_pkTcpSession->Init (shared_from_this ());
}

void CTcpConnection::Shutdown ()
{
	m_pkNetBridge = nullptr;

	m_pkTcpSession->Shutdown ();
	m_pkTcpSession = nullptr;
}

void CTcpConnection::OnDisconnect ()
{
	m_pkNetBridge->OnDisconnect ();
}

void CTcpConnection::ResolveInput (CBitInStream& _rkInStream)
{
	unsigned short protocolID;
	_rkInStream.Read (protocolID);

	std::shared_ptr<INetProtocol> protocol = CProtocolManager::GenerateProtocol (protocolID);
	if (protocol != nullptr)
	{
		protocol->SetNetBridge (m_pkNetBridge);
		protocol->Deserialize (_rkInStream);
		protocol->Excute ();
	}
}

void CTcpConnection::ComposeOutput (std::shared_ptr<INetProtocol> _pkProtocol)
{
	CBitOutStream outStream;
	_pkProtocol->OnSerialize (outStream);
	m_pkTcpSession->Write (outStream);
}

File: Unity/Project Nothing/Assets/Script/Network/TcpConnection.cs
using System.Collections;
using System.Collections.Concurrent;
using System.Net;

namespace ProjectNothing
{
    public sealed class TcpConnection
    {
        readonly TcpSession m_TcpSession = new ();
        readonly ConcurrentQueue<INetProtocol> m_QueuedProtocols = new ();

        bool m_IsInit = false;

        public IEnumerator Init (IPAddress ipAddress, int port)
        {
            yield return m_TcpSession.Init (this, ipAddress, port);

            m_IsInit = true;
        }

        public void Shutdown ()
        {
            m_TcpSession.Shutdown ();
        }

        public void Update ()
        {
            if (!m_IsInit)
            {
                return;
            }

            foreach (INetProtocol protocol in m_QueuedProtocols)
            {
                protocol.Excute ();
            }

            m_QueuedProtocols.Clear ();
        }

        public void ResolveInput (BitInStream inStream)
        {
            inStream.Read (out ushort protocolID);

            INetProtocol protocol = ProtocolManager.GenerateProtocol (protocolID);
            if (protocol != null)
            {
                protocol.Deserialize (inStream);
                m_QueuedProtocols.Enqueue (protocol);
            }
        }

        public void ComposeOutput (INetProtocol protocol)
        {
            BitOutStream outStream = new ();
            protocol.OnSerialize (outStream);
            m_TcpSession.Write (outStream);
        }
    }
}